src/main/java/io/hops/hopsworks/expat/migrations/featurestore/statistics/StatisticsMigration.java

```diff
@@ -0,0 +1,691 @@
+  
+  private final static String FEATURE_GROUP = "FEATURE_GROUP";
+  private final static String TRAINING_DATASET = "TRAINING_DATASET";
+  
+  public StatisticsMigration() {
+  }
+  
+  @Override
+  public void migrate() throws MigrationException {
+    LOGGER.info("Starting migration of " + super.getClass().getName());
+    
+    try {
+      setup();
+      runMigration();
+      close();
+    } catch (ConfigurationException | SQLException ex) {
+      String errorMsg = "Could not initialize database connection";
+      LOGGER.error(errorMsg);
+      close();
+      throw new MigrationException(errorMsg, ex);
+    } catch (IOException | IllegalAccessException | InstantiationException ex) {
+      String errorMsg = "Could not migrate statistics";
+      LOGGER.error(errorMsg);
+      throw new MigrationException(errorMsg, ex);
+    }
+    LOGGER.info("Finished migration of " + super.getClass().getName());
+  }
+  
+  @Override
+  public void rollback() throws RollbackException {
+    LOGGER.info("Starting rollback of " + super.getClass().getName());
+    try {
+      setup();
+      runRollback();
+      close();
+    } catch (ConfigurationException | SQLException ex) {
+      String errorMsg = "Could not initialize database connection";
+      LOGGER.error(errorMsg);
+      close();
+      throw new RollbackException(errorMsg, ex);
+    }
+    LOGGER.info("Finished rollback of " + super.toString());
+  }
+  
+  public void runMigration()
+    throws MigrationException, SQLException, IOException, IllegalAccessException, InstantiationException {
+    PreparedStatement fdsStmt = null;
+    PreparedStatement orphanStatsStmt = null;
+    
+    try {
+      connection.setAutoCommit(false);
+      
+      // migrate feature descriptive statistics:
+      // - insert statistics (per feature) into feature descriptive statistics table
+      // - insert intermediate fg/td descriptive statistics rows
+      // - update/remove statistics file. Only histograms, correlations, kll and unique values are kept in the file
+      // - update start_commit_window in affected feature group statistic rows
+      fdsStmt = connection.prepareStatement(GET_FEATURE_DESCRIPTIVE_STATISTICS);
+      ResultSet fdsResultSet = fdsStmt.executeQuery();
+      migrateFeatureDescriptiveStatistics(fdsResultSet);
+      fdsStmt.close();
+      
+      // delete orphan statistics files. There are two possible reasons why stats file are orphan during the migration:
+      // - if multiple FG statistics on the same commit id, only the most recent is migrated, the rest become orphan.
+      // - if multiple TD statistics on the same dataset, only the most recent is migrated, the rest become orphan.
+      orphanStatsStmt = connection.prepareStatement(GET_ORPHAN_STATISTICS);
+      ResultSet orphanStatsResultSet = orphanStatsStmt.executeQuery();
+      deleteOrphanStatisticsFiles(orphanStatsResultSet);
+      orphanStatsStmt.close();
+      
+      connection.commit();
```

Comment:
This might be an issue I think if there are a lot of statistics... might consider batching?

Reply:
I don't expect a lot of orphan statistics here. We have just fixed a case resulting in orphan statistics in 3.3 and 3.4.
Regarding the two cases mentioned in the comment, those statistics on the same commit id or same TD are triggered manually by the user either using "compute statistics" in the UI or .compute_statistics() in hsfs. These are not common user actions I think, so I wouldn't expect a lot of orphan statistics.